lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,235 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using Newtonsoft.Json;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Common.Telemetry
+{
+    /// <summary>
+    /// MetricData class maintains the current value of a single metric and keeps count of the number
+    /// of times this metric has been updated. If the metric is immutable, it keeps a record of updates.
+    /// Once the data has been processed, the records and count will reset.
+    /// </summary>
+    [JsonObject]
+    public sealed class MetricTracker : ITracker
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricTracker));
+
+        [JsonProperty]
+        internal IMetric Metric;
+
+        [JsonProperty]
+        internal bool KeepUpdateHistory;
+
+        /// <summary>
+        /// if KeepUpdateHistory is true, keeps a history of updates.
+        /// </summary>
+        [JsonProperty]
+        internal ConcurrentQueue<MetricRecord> Records;
+        
+        /// <summary>
+        /// Number of times metric has been updated since last processed.
+        /// </summary>
+        [JsonProperty]
+        internal int ChangesSinceLastSink;
+
+        private IDisposable _unsubscriber;
+
+        /// <summary>
+        /// Constructor for metricData
+        /// </summary>
+        /// <param name="metric"></param>
+        /// <param name="initialValue"></param>
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            ChangesSinceLastSink = 0;
+            KeepUpdateHistory = metric.IsImmutable;
+            Records = new ConcurrentQueue<MetricRecord>();
+            if (KeepUpdateHistory)
+            {
+                Records.Enqueue(CreateMetricRecord(metric));
+            }
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IMetric metric, int changesSinceLastSink, ConcurrentQueue<MetricRecord> records, bool keepUpdateHistory)
+        {
+            Metric = metric;
+            Records = records;
+            KeepUpdateHistory = keepUpdateHistory;
+            ChangesSinceLastSink = changesSinceLastSink;
+        }
+
+        /// <summary>
+        /// Flush records currently held in the records queue.
+        /// </summary>
+        /// <returns>A queue containing all the flushed records.</returns>
+        internal ConcurrentQueue<MetricRecord> FlushChangesSinceLastSink()
+        {
+            ConcurrentQueue<MetricRecord> records = new ConcurrentQueue<MetricRecord>();
+            if(!Records.IsEmpty)
```

Comment:
Originally Records was a private field of the class, and the way to access it was through GetMetricRecords that would either return the private _records OR return a queue with a single record. In other words, when not keeping history, private _records will always be empty, but when someone asks for the metric records through the api, they will always get at least one value.
Then _records became internal Records to avoid messy casing in the json constructor, but if we are ok with naming private fields in Pascal case then I think logically it would be clearer.

Reply:
Keeping the Records.IsEmpty check for now because if GetMetricRecords() were used, an additional ConcurrentQueue would be allocated when the Records field is empty.